🚨 🚨 🚨 Fix Issue 15003: SentencePiece Tokenizers Not Adding Special Tokens in `convert_tokens_to_string`

beneyal opened the pull request:
Hello! 👋 

This PR fixes issue #15003 and is also related to issue #11716.

First, I've expanded the `test_sentencepiece_tokenize_and_convert_tokens_to_string` test in `test_tokenization_common.py` which, in addition to checking a round-trip from string to tokens and back, performs a check whether the tokens corresponding to `input_ids` (using `convert_ids_to_tokens`) followed by `convert_tokens_to_string` and then re-tokenized, give the same number of tokens as `input_ids`. This is because not all SentencePIece tokenizers have a `*Fast` flavor, and so I couldn't compare behaviors similar to the `T5Tokenizer` test.

This test failed for the following tokenizers:
- `AlbertTokenizer`
- `BarthezTokenizer`
- `CamembertTokenizer`
- `DebertaV2Tokenizer`
- `FNetTokenizer`
- `LayoutXLMTokenizer`
- `M2M100Tokenizer`
- `MarianTokenizer`
- `MBart50Tokenizer`
- `PegasusTokenizer`
- `Speech2TextTokenizer`


Most of these tokenizers had the same implementation for `convert_tokens_to_string`:

```python
def convert_tokens_to_string(self, tokens):
    return self.sp_model.decode(tokens)
```

All tokenizers who had this implementation, or something that can be reduced to it, are now inheriting `SentencePieceStringConversionMixin` (defined in `tokenization_utils.py`), and the implementation uses the same code as PR #8435. The rest have a variation of this code, and perhaps a future refactor can DRY it a bit.

The only tokenizer that doesn't pass the test, and is therefore excluded from it, is `LayoutXLMTokenizer`, whose `convert_tokens_to_string` method doesn't use the SentencePiece model in its implementation, and so I figured the lack of special tokens is intentional, rather than a byproduct of SentencePiece like the rest.

I think this PR is relevant for @NielsRogge who initially opened the issue on `AlbertTokenizer`, and following the PR guidelines, I think this is also relevant to @LysandreJik and @patrickvonplaten.

This has been a very fun project, and hopefully you will allow me to contribute more in the future 🙂 

Thanks,
Ben

beneyal committed:
Add test for SentencePiece not adding special tokens to strings

beneyal committed:
Add SentencePieceStringConversionMixin to fix issue 15003

beneyal committed:
Fix conversion from tokens to string for most SentencePiece tokenizers

Tokenizers fixed:
- AlbertTokenizer
- BarthezTokenizer
- CamembertTokenizer
- FNetTokenizer
- M2M100Tokenizer
- MBart50Tokenizer
- PegasusTokenizer
- Speech2TextTokenizer

beneyal committed:
Fix MarianTokenizer, adjust SentencePiece test to accomodate vocab

beneyal committed:
Fix DebertaV2Tokenizer

beneyal committed:
Ignore LayoutXLMTokenizer in SentencePiece string conversion test

beneyal committed:
Run 'make style' and 'make quality'

HuggingFaceDocBuilder commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_15775). All of your documentation changes will be reflected on that endpoint.

SaulLu commented on tests/test_tokenization_common.py:
```diff
@@ -356,7 +357,7 @@ def test_sentencepiece_tokenize_and_convert_tokens_to_string(self):
 
         tokenizer = self.get_tokenizer()
-        text = "This is text to test the tokenizer."
+        text = "This is a test"
```
Is there a particular reason why you changed this example? :slightly_smiling_face:

beneyal replied:
Yes, it's because of the test of `MarianTokenizer` uses a very specific and small vocabulary, as shown in the snippet, so I thought it wouldn't make much of a difference to the other tokenizers if I made that change. Without the change, `MarianTokenizer` fails the test.

https://github.com/huggingface/transformers/blob/35ecf99cc4e1046e5908f5933a4548e43a0f4f1d/tests/marian/test_tokenization_marian.py#L57

SaulLu replied:
I am surprised because I tested your test by changing only this value and I do not encounter any errors for MarianTokenizer. Could you please re-test and share the error you had? :pray:

beneyal replied:
Well, that's embarrassing. I know `MarianTokenizer` gave me an error before, but now it no longer does. I reverted the `text` variable to be the original string.

SaulLu commented on tests/test_tokenization_common.py:
```diff
@@ -368,11 +369,21 @@ def test_sentencepiece_tokenize_and_convert_tokens_to_string(self):
-        if self.test_sentencepiece_ignore_case:
-            reverse_text = reverse_text.lower()
+        # All tokenizers pass this test without the below commented out code.
+        # if self.test_sentencepiece_ignore_case:
+        #     reverse_text = reverse_text.lower()
```
I just tested it locally, and I have no errors when uncommenting these lines. Do you specifically remember the error that caused you to comment out these lines?

SaulLu replied:
Oh I understand! I think it's because currently no tokenizer has the lowercasing argument enabled. I'd be in favor of leaving it as is in case a future tokenizer uses this setup

beneyal replied:
No problem, I reverted the change.

SaulLu commented on tests/test_tokenization_common.py:
```diff
@@ -368,11 +369,21 @@ def test_sentencepiece_tokenize_and_convert_tokens_to_string(self):
 
         self.assertEqual(reverse_text, text)
 
+        if isinstance(tokenizer, LayoutXLMTokenizer):
+            return
```
Just for consistency with the rest of the code, when a `test_tokenization_common.py` test does not apply to a particular tokenizer, what we do is override that test in its test file and comment out why the test does not apply to it.

See for example what has been done for ByT5:

https://github.com/huggingface/transformers/blob/35ecf99cc4e1046e5908f5933a4548e43a0f4f1d/tests/byt5/test_tokenization_byt5.py#L309-L311

Would you mind making this change? :relaxed:

beneyal replied:
Gladly!

SaulLu replied:
Thank you!!

SaulLu commented on tests/test_tokenization_common.py:
```diff
@@ -368,11 +369,21 @@ def test_sentencepiece_tokenize_and_convert_tokens_to_string(self):
+        if isinstance(tokenizer, LayoutXLMTokenizer):
+            return
+
+        input_ids = tokenizer(text).input_ids
+        tokens_including_special = tokenizer.convert_ids_to_tokens(input_ids)
+        reverse_text = tokenizer.convert_tokens_to_string(tokens_including_special)
+
+        self.assertEqual(len(tokenizer.tokenize(reverse_text)), len(input_ids))
```
If I understood correctly, the initial problem was that the special token was not found in the `convert_tokens_to_string` output while given as individual tokens in the input. So, I think we should refocus the test on this check by testing the method on all the special tokens of each spm tokenizer - which are stored in the `all_special_tokens` attribute.

So I suggest the following changes:
```suggestion
        special_tokens =  tokenizer.all_special_tokens
        special_tokens_string = tokenizer.convert_tokens_to_string(special_tokens)
        for special_token in special_tokens:
            self.assertIn(special_token, special_tokens_string)
```

These changes cause the test to fail now for `ReformerTokenizer`, `BigBirdTokenizer` and `BertGenerationTokenizer` which are also spm tokenizers. 

Don't hesitate to tell me if you ever wrote the previous test with another case in mind! :smile:

beneyal replied:
Great idea! This test is definitely more robust than my length comparison, especially it it managed to catch a few more SPM tokenizers.

SaulLu commented on tests/test_tokenization_common.py:
```diff
@@ -368,11 +369,21 @@ def test_sentencepiece_tokenize_and_convert_tokens_to_string(self):
+        reverse_text = tokenizer.convert_tokens_to_string(tokens_including_special)
+
+        self.assertEqual(len(tokenizer.tokenize(reverse_text)), len(input_ids))
+
```
What I don't necessarily know how to handle in this `convert_tokens_to_string` function is how to re-add spaces around each special token. I wonder if we should be aligned with what the fast version of the tokenizer would outputs. If it's something that seems relevant to you, I would add this test:
```suggestion

        if self.test_rust_tokenizer:
            rust_tokenizer = self.get_rust_tokenizer()
            special_tokens_string_rust = rust_tokenizer.convert_tokens_to_string(special_tokens)
            self.assertEqual(special_tokens_string, special_tokens_string_rust)
```

beneyal replied:
Initially, I didn't go for the "compare to Rust tokenizer" since not all tokenizers in question have this luxury. Thinking about it now that you mention it, this code will definitely add another layer of reliability to the existing test.

beneyal replied:
Okay, after adding this test, a few previously passing tests failed. This is because of the space you mentioned in another comment. I removed the space, so now the code looks like this:

```python
def convert_tokens_to_string(self, tokens):
    ...
    out_string += self.sp_model.decode(current_sub_tokens) + token  # << a space used to be here
    ...
```

And now all the tests pass, including `T5Tokenizer`.
I will note that I changed calls to `self.sp_model.decode_pieces` to `self.sp_model.decode`, as `decode`, if needed, delegates to `decode_pieces`, and this makes the code uniform, in my opinion.

SaulLu commented on src/transformers/tokenization_utils.py:
```diff
@@ -956,3 +956,19 @@ def _decode(
             return text
+
+
+class SentencePieceStringConversionMixin:
```
I completely understand why you proposed to create a `mixin` class for this method common to many tokenizers. It seems that for the moment the philosophy of the transformers library is rather to duplicate the code than to make inheritances.

I would love to know the opinion of the `transformers` maintainers on this question. @LysandreJik , @sgugger , @patrickvonplaten  or @patil-suraj what you think about this design?

sgugger replied:
For a short method like this, I would advise to put it in every tokenization file. Users don't like to have to read superclasses code and like that all the code is available in one place, so it's a core principle of our philosophy to have all code in one single file for the obecets of the library.

beneyal replied:
No problem, I'm all in on that philosophy, especially since there are, I think, 3 special cases where I had to almost duplicate the code anyway.

For context, I decided on the mixin approach to try and mimic the thought process behind `GenerationMixin` and other such `*Mixin` classes I stumbled upon throughout the codebase exploration.

SaulLu replied:
Great :smile:, thank you so much!

SaulLu commented on src/transformers/tokenization_utils.py:
```diff
@@ -956,3 +956,19 @@ def _decode(
+        for token in tokens:
+            # make sure that special tokens are not decoded using sentencepiece model
+            if token in self.all_special_tokens:
+                out_string += self.sp_model.decode(current_sub_tokens) + token + " "
```
This is related to my previous comment, but I'm not sure what the behavior should be here to add spaces. How did you choose this strategy?

beneyal replied:
This is the exact code from PR #8435, I assumed this implementation would be relevant for all SPM tokenizers, but it may well be that this assumption is incorrect.

beneyal replied:
After running the test you suggested, I do believe that this extra space is an error, at least compared to the Rust tokenizers. Removing the space passes the test 👍

SaulLu replied:
I won't be able to look at this issue `T5Tokenizer` with the again until next Monday but I promise to come back to you early next week

beneyal replied:
Thank you very much! I will be waiting patiently 😄

SaulLu replied:
Actually this error with T5 makes me think that we probably have the same problem with the other tokenizers. Would you like to add a new test on `decode` that would check this?
 
I think that the intrication between `decode` and `convert_tokens_to_string` is very subtle and we need to be very careful about these behaviors. :smile:

beneyal replied:
Will do!

SaulLu replied:
Thank you 🤩

beneyal replied:
I have written the following test:
```python
def test_sentencepiece_tokenize_and_decode(self):
    if not self.test_sentencepiece:
        return

    text = "This is text to test the tokenizer."
    if self.test_rust_tokenizer:
        tokenizer = self.get_tokenizer()
        rust_tokenizer = self.get_rust_tokenizer()

        slow_ids = tokenizer(text).input_ids
        fast_ids = rust_tokenizer(text).input_ids
        self.assertEqual(slow_ids, fast_ids)

        slow_decoded = tokenizer.decode(slow_ids)
        fast_decoded = rust_tokenizer.decode(slow_ids)
        self.assertEqual(slow_decoded, fast_decoded)
```

The behaviors of slow and fast tokenizers in `decode` are indeed mismatched, and the mismatch is, to me, quite odd. Let's take a look at a few failures (these tokenizers pass `test_sentencepiece_tokenize_and_convert_tokens_to_string`):

|Tokenizer|Output|
|-|-|
|Albert|`'[CLS]this is text to test the tokenizer.[SEP]' != '[CLS] this is text to test the tokenizer.[SEP]'`|
|Barthez|`'<s>This is text to test the tokenizer.</s>' != '<s> This is text to test the tokenizer.</s>'`|
|BigBird|`'[CLS] This is text to test the tokenizer. [SEP]' != '[CLS] This is text to test the tokenizer.[SEP]'`|
|Camembert|`'<s>This is text to test the tokenizer.</s>' != '<s> This is text to test the tokenizer.</s>'`|
|MBart50|`'en_XXThis is text to test the tokenizer.</s>' != 'en_XX This is text to test the tokenizer.</s>'`|

It would seem that the correct behavior, as dictated by the fast tokenizers, is that the BOS token should have a space after it, but the EOS token comes right after the last token without a space.

This seems like a small thing to fix, but the question is whether the problem is that space, or is it something bigger that I'm not aware of.

beneyal replied:
Hi! Quick update: I added the test for `decode` (could really use your input on this one 🙂) and fixed the tokenizers from the above table + some others. Most required a quick fix (basically: don't add a space if the previous token was a special token); BigBird, FNet, and XLNet required an override of `_decode` in addition to the fix that worked for the rest.

Your feedback is welcome whole-heartedly,  and have an excellent weekend!

beneyal replied:
Hi @SaulLu 🙂 

I hope everything is okay with you, and that I'm not bothering you by tagging. It has been a while, so I'm just checking this PR doesn't "fall between the chairs" (I'm not even sure people say this in English 😅).

Have a wonderful day and I'm eagerly awaiting your reply!

beneyal replied:
No problem, have a great weekend and we'll pick this up next week! 🙂

SaulLu replied:
So I really spent a lot of time thinking about your PR. I'm already impressed with everything you've done! :clap: 

The only thing that bothers me at the moment is the fact that we're including a lot of complexity in this RP to solve a problem (the output between the slow and the fast version is not the same) that is not the problem we wanted to solve initially. This second problem is in fact quite complex - beyond the case of sentencepiece-based tokenizers - and we are discussing it "internally" to see what we want to do (in the short and long term). So I'm sorry again, but I think it would be really beneficial to wait a bit longer to get more feedback. :disappointed: 

Regarding the initial problem, I was wondering if you managed to identify why the sentencepiece decode method ignored some tokens. Is it only the BOS, EOS and PAD type tokens?

beneyal replied:
> Regarding the initial problem, I was wondering if you managed to identify why the sentencepiece decode method ignored some tokens. Is it only the BOS, EOS and PAD type tokens?

I haven't, as I didn't think we want to delve in the C++ implementation of SentencePiece. From what I understood, the code we added was only circumventing a problem with SP without touching the external dependency.

I have no problem waiting for further feedback, this PR and myself aren't planning on going anywhere anytime soon.

Looking forward to continue working with you, and I'm thankful to have you as my reviewer 🙂

SaulLu replied:
Thank you very much for your understanding!

> I haven't, as I didn't think we want to delve in the C++ implementation of SentencePiece. From what I understood, the code we added was only circumventing a problem with SP without touching the external dependency.

Yes, you're absolutely right, we are not going to touch the sentencepiece library at all, but it was more to make sure that there are no cases that we could miss. My guess is that sentencepiece model is just ignoring the bos_piece, eos_piece and pad_piece (listed [here](https://github.com/google/sentencepiece/blob/master/doc/options.md)) when decoding but I'm not 100% sure :smile:

beneyal replied:
Hey @SaulLu 🙂 

Hope you're doing well and keeping healthy!

Since the issue is now marked as stale, I thought I'll check the status of this issue with you.

Have an excellent week!

beneyal replied:
Hi @SaulLu,

How are you?

Any news regarding this issue?

Thanks 🙂

beneyal replied:
I think we should split the decision in two:
1. Tokenizers that don't have a fast version, and then there's nothing to compare to, so users using `AutoTokenizer.from_pretrained()` even with `use_fast=True` will get a slow tokenizer
2. Tokenizers that have both versions. This part, I think, will require some investigation from HF to see whether there are use-cases of a slow tokenizer when a fast one is available. As a user, I do `pip install transformers` and get `tokenizers` as a dependency. Then `AutoTokenizer.from_pretrained()` by default gives a fast tokenizer, if one's available. Why would I want to use a slow tokenizer? Answering this question will motivate the decision on what to do with the slow tokenizers' code. It could be that the code can be ugly-but-working since it won't get touched too often, but if the fast tokenizers get changed frequently and we want to keep feature parity with the slow tokenizers, then maintainable code is probably a priority.

All the best, Lucile 🙂

SaulLu replied:
Hi @beneyal ,

As promised I've spent some time today diving back into your PR. In the end, I think that what you have proposed is a good solution given all the constraints that we have.

>  Why would I want to use a slow tokenizer?
To come back to one of the last questions you raised, we have slow versions for several reasons, including the fact that they are extremely close to the original code of the authors (this is the source of truth for building the fast tokenizer - however this source of truth disappears when some features have not been implemented in the original code, which is very often the case with added tokens), they are python versions so they are more readable by users and for backward compatibility reasons (e.g. there are options in the `decode` method of the slow tokenizers that are not available in the fasts). 

Before I do a new PR review, do you think you'll have time to resolve the merge conflicts? If you don't have the time, don't hesitate to tell me, I can take care of it and open a PR on your branch with the necessary changes. :smile: 

Thank you again for your patience and for having worked so hard on the code base!

beneyal replied:
Hey @SaulLu,

I've fixed the merge conflicts and running all the tokenization tests right now (takes a while 😅), and I see that DeBERTa v2 is failing. Should I try and fix it?

Thank ***you*** for your tremendous help 🙂

SaulLu replied:
Thanks again for your responsiveness! Regarding DeBERTa v2, could you share with me the tests that fail?

beneyal replied:
All of them 😅 Once the testing finishes (currently 50%), I'll know exactly why. For context, I accepted the changes in `main` over what was in the file of my branch. Didn't do anything else with that file.

![image](https://user-images.githubusercontent.com/3891274/165714381-556fa947-2f65-43dc-a86c-542c355e0cae.png)

beneyal replied:
So, after feeling dumb for not noticing the `pytest -n auto` argument, which is why the tests took almost 3 hours to run, I found the problem: when accepting the change, it removed a constructor argument I passed. I fixed that, and now only one test is failing:

![image](https://user-images.githubusercontent.com/3891274/165733745-7fa5ce23-ff54-4aa8-a621-f900dbb54171.png)

This is weird, since using the "microsoft/deberta-v2-xlarge" tokenizer works on both slow and fast. I believe the problem is the test vocab. Is there something I can do to fix it, other than changing the test sentence (which seems quite drastic)?

beneyal replied:
Interesting. This change breaks `test_get_vocab`, which makes sense, but `test_sentencepiece_tokenize_and_decode` at least now shows the "real" problem, which is what I fixed for the other slow tokenizers: 

```
 AssertionError: '[CLS]<unk>his is text to test the tokenizer.[SEP]' != '[CLS] <unk>his is text to test the tokenizer.[SEP]'
```

It's the spacing thing. For some tokenizers, it was a simple fix, for others, it required overriding `_decode`, but I can fix this. I'll need to check how the fast tokenizer works to mimic it space-between-special-tokens-wise.

beneyal replied:
Hi @SaulLu,

I made DeBERTa v2 act like the fast tokenizer. It's very ad-hoc, since the behavior of the special tokens is a bit weird, but it passes all the tests 🙂 

Would love to hear your feedback!

SaulLu commented on src/transformers/models/pegasus/tokenization_pegasus.py:
```diff
@@ -226,10 +226,10 @@ def _convert_id_to_token(self, index: int) -> str:
-        """Converts a sequence of tokens (string) in a single string."""
-        out_string = self.sp_model.decode_pieces(tokens)
-        return out_string
+    # def convert_tokens_to_string(self, tokens):
+    #     """Converts a sequence of tokens (string) in a single string."""
+    #     out_string = self.sp_model.decode_pieces(tokens)
+    #     return out_string
```
Nit, I think it fell through the cracks :smile:

beneyal replied:
Definitely did! 😅

SaulLu commented on src/transformers/models/mbart50/tokenization_mbart50.py:
```diff
@@ -233,9 +233,9 @@ def _convert_id_to_token(self, index: int) -> str:
-    def convert_tokens_to_string(self, tokens: List[str]) -> str:
-        """Converts a sequence of tokens (strings for sub-words) in a single string."""
-        return self.sp_model.decode(tokens)
+    # def convert_tokens_to_string(self, tokens: List[str]) -> str:
+    #     """Converts a sequence of tokens (strings for sub-words) in a single string."""
+    #     return self.sp_model.decode(tokens)
```
Nit, too :)

beneyal replied:
Oops again :)

SaulLu reviewed (commented):
Thank you so much for working to resolve this issue! It was indeed a behavior that was troublesome! :smile: 

I also really appreciate that you put a test in your first pull review request! 

I think one of the difficulties of the problem you are attacking is how to readjust spaces around these special tokens.  Thinking about it quickly, I think the result should be identical to the one returned by the rust version of the tokenizer.

beneyal committed:
Clean convert_tokens_to_string test

Instead of explicitly ignoring LayoutXLMTokenizer in the test,
override the test in LayoutLMTokenizationTest and do nothing in it.

beneyal committed:
Remove commented out code

beneyal committed:
Improve robustness of convert_tokens_to_string test

Instead of comparing lengths of re-tokenized text and input_ids,
check that converting all special tokens to string yields a string
with all special tokens.

beneyal committed:
Inline and remove SentencePieceStringConversionMixin

The convert_tokens_to_string method is now implemented
in each relevant SentencePiece tokenizer.

beneyal committed:
Run 'make style' and 'make quality'

beneyal committed:
Revert removal of space in convert_tokens_to_string

Commit:
Remove redundant import

Commit:
Revert test text to original

Commit:
Uncomment the lowercasing of the reverse_text variable

beneyal committed:
Mimic Rust tokenizer behavior for tokenizers

- Albert
- Barthez
- Camembert
- MBart50
- T5

beneyal committed:
Fix accidentally skipping test in wrong tokenizer

beneyal committed:
Add test for equivalent Rust and slow tokenizer behavior

beneyal committed:
Override _decode in BigBirdTokenizer to mimic Rust behavior

beneyal committed:
Override _decode in FNetTokenizer to mimic Rust behavior

beneyal committed:
Override _decode in XLNetTokenizer to mimic Rust behavior

beneyal committed:
Merge 'main' into the 15003 fix branch

beneyal committed:
Remove unused 're' import

beneyal committed:
Update DebertaV2Tokenizer to mimic Rust tokenizer

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_15775). All of your documentation changes will be reflected on that endpoint.

SaulLu commented on src/transformers/models/bert_generation/tokenization_bert_generation.py:
```diff
@@ -149,8 +149,17 @@ def _convert_id_to_token(self, index):
 
     def convert_tokens_to_string(self, tokens):
         """Converts a sequence of tokens (string) in a single string."""
-        out_string = self.sp_model.decode_pieces(tokens)
-        return out_string
+        current_sub_tokens = []
+        out_string = ""
+        for token in tokens:
+            # make sure that special tokens are not decoded using sentencepiece model
+            if token in self.all_special_tokens:
+                out_string += self.sp_model.decode(current_sub_tokens) + token
+                current_sub_tokens = []
+            else:
+                current_sub_tokens.append(token)
+        out_string += self.sp_model.decode(current_sub_tokens)
+        return out_string.strip()
```
I would propose to take the same solution as the tokenizer with a fast version because if we add a fast version later we will be blocked - and the decoding method is quite important for Bert generation.

beneyal replied:
You mean like, for example, the `convert_tokens_to_string` method in `AlbertTokenizer`?

SaulLu commented on src/transformers/models/deberta_v2/tokenization_deberta_v2.py:
```diff
@@ -336,7 +341,27 @@ def convert_ids_to_tokens(self, ids):
+                    out_string += self.spm.decode_pieces(current_sub_tokens) + token
+                    prev_is_special = True
+                    current_sub_tokens = []
+                    if token == self.special_tokens[2]:  # if token == <unk>
+                        seen_unk = True
```
I'm going to sound annoying but I think it's risky to do it this way because there's nothing that requires the unknown token to be in the third position. Do you remember if another tokenizer is in the same situation?

beneyal replied:
No other tokenizer had this issue during my work on this PR. I agree it's risky, and it's not even aligned with the comment that appears in the `__init__` method. That's the problem with being an "outsider", I don't know if there's a mistake in the fast tokenizer or am I missing something 😅 The behavior of `<unk>` is different in DeBERTa v2 than in the other SP tokenizers.

sgugger replied:
This special code should indeed be removed. I think it's okay to exclude Deberta-v2 from the new test, decode as it makes the most sense even if it breaks the compatibility with the fast tokenizer if we think there is a bug there. We don't guarantee 100% match between the two in all situations.

beneyal commented:
GitHub Actions, stop marking this PR as stale! 😅 

Hi @SaulLu, how are you? I started a new job so kind of lost track with this PR, is it still relevant? I'd really like to see this PR merged at some point 😁

beneyal commented:
Hi again @SaulLu 🙂 

I found something interesting.

Running this code:

```python
from transformers import AutoTokenizer

tokenizers = [
    (
        "microsoft/deberta-v2-xlarge",
        AutoTokenizer.from_pretrained("microsoft/deberta-v2-xlarge", use_fast=True),
        AutoTokenizer.from_pretrained("microsoft/deberta-v2-xlarge", use_fast=False),
    ),
    (
        "albert-base-v2",
        AutoTokenizer.from_pretrained("albert-base-v2", use_fast=True),
        AutoTokenizer.from_pretrained("albert-base-v2", use_fast=False),
    ),
    (
        "bert-large-uncased",
        AutoTokenizer.from_pretrained("bert-large-uncased", use_fast=True),
        AutoTokenizer.from_pretrained("bert-large-uncased", use_fast=False),
    ),
    (
        "camembert-base",
        AutoTokenizer.from_pretrained("camembert-base", use_fast=True),
        AutoTokenizer.from_pretrained("camembert-base", use_fast=False),
    ),
]

text = "This is some {} to the tokenizer."

print("| Checkpoint Name | Fast Tokenizer | Slow Tokenizer |")

for ckpt, fast, slow in tokenizers:
    fast_tokens = fast.tokenize(text.format(fast.unk_token))
    slow_tokens = slow.tokenize(text.format(slow.unk_token))
    assert (
        fast_tokens == slow_tokens
    ), f"tokens should be identical in the same model. {ckpt = }, {slow_tokens =}, {fast_tokens = }"
    fast_string = fast.convert_tokens_to_string(fast_tokens)
    slow_string = slow.convert_tokens_to_string(fast_tokens)
    print(f"| {ckpt} | {fast_string} | {slow_string} |")
```

Gives us the following table:

| Checkpoint Name | Fast Tokenizer | Slow Tokenizer |
|---|---|---|
| microsoft/deberta-v2-xlarge | This is some[UNK] to the tokenizer. | This is some[UNK]to the tokenizer. |
| albert-base-v2 | this is some\<unk\> to the tokenizer. | this is some\<unk\>to the tokenizer. |
| bert-large-uncased | this is some [UNK] to the tokenizer. | this is some [UNK] to the tokenizer . |
| camembert-base | This is some\<unk\> to the tokenizer. | This is some\<unk\>to the tokenizer. |


So we can see that DeBERTa v2, Albert and CamemBERT have the same behavior w.r.t UNK:
* Fast tokenizer has no space between a non-special token and UNK
* Slow tokenizer has no spaces on both sides of the UNK

That's weird, because then we would expect DeBERTa v2's `convert_tokens_to_string` code (which I made identical to Albert's) to pass the tests we wrote.

So I made a slight change to the test (`test_tokenization_deberta_v2.py:40`):

```python
# tokenizer = DebertaV2Tokenizer(SAMPLE_VOCAB, unk_token="<unk>")
tokenizer = DebertaV2Tokenizer.from_pretrained(
    "https://huggingface.co/microsoft/deberta-v2-xlarge/resolve/main/spm.model"
)
```

Now the `convert_tokens_to_string` tests pass, and the only failing tests are ones that rely on the size of the test vocabulary and some casing (`<pad>` vs `<PAD>`). Add this to the fact the only DeBERTa v2's test needed to specify a different UNK token, I believe the logic is fine, but something is amiss with the test vocab. What do you think?

Have a lovely weekend!

beneyal commented:
Hello @SaulLu 🙂 Just a little ping in case you didn't see my written novel of a comment above 😅 Thank you again for your time and have a lovely day!

beneyal commented:
Hi @SaulLu, GitHub Actions wants to close this issue again 😅 Hope everything is well!

I haven't worked on this since June 11th because I'm waiting for your insights 🙂 

Have a great week!

SaulLu commented:
Hi @beneyal , 

I wanted to leave you a message to let you know that I haven't had time so far to dive into your PR (and I, unfortunately, have no idea what the right answer is by reading your findings) and that unfortunately, I'm going to be off next week.

I appreciate a lot all the efforts you made for this fix, I really hope we will find a satisfactory solution!

beneyal commented:
Hi @SaulLu 🙂 Any news? The bot wants to close the issue again 😅

beneyal commented:
Hi @SaulLu, how are you? 🙂 

Anything new regarding this PR? 🙏

beneyal commented:
This PR is still in progress 🙂

dvfabbri commented:
The Albert changes resolved the lack of `[MASK]` for me. Plans to get this merged?

sgugger commented on tests/test_tokenization_common.py:
```diff
@@ -369,11 +369,39 @@ def test_sentencepiece_tokenize_and_convert_tokens_to_string(self):
         # check if converting back to original text works
         reverse_text = tokenizer.convert_tokens_to_string(tokens)
 
+        # All tokenizers pass this test without the below commented out code.
```
```suggestion
        # All tokenizers pass this test without the below commented out code.
```
To clean up?

beneyal replied:
I'll remove the comment 🙂

sgugger reviewed (commented):
Sorry this PR has been left stale for so long @beneyal. I'll take over the reviews from now on as Lucile has been super busy on other projects!

First of all, this PR contains some breaking changes (for the better). Could you add three 🚨 in the title so that when we draft the next release notes, we clearly see it and can mention it in our section about breaking changes that fix standing bugs?

Then I think the most standing comment was on the Deberta-v2 tokenizer.

beneyal commented:
@sgugger Wonderful to have you on board! 🚀 

I can't comment on the Deberta snippet, so I'll try here: we basically want the code for Deberta V2 be the same as the code for the other tokenizers like Albert, i.e., without the special `seen_unk` hack?

I also changed the title as requested 🙂

sgugger commented:
Yes, that would be great! And if you could fix the merge conflicts/rebase on main that would also be awesome!

beneyal commented:
@sgugger I have a rather silly question: I see in GitHub there are two conflicts: `tests/deberta_v2/test_tokenization_deberta_v2.py` and `tests/layoutxlm/test_tokenization_layoutxlm.py`. When merging `upstream/main` to my branch, there are no conflicts. Regarding rebasing, I've never done it so I'm not familiar with that functionality and what do I rebase into what. Could you please help me with this? 😅

beneyal committed:
Merge branch 'main' of https://github.com/huggingface/transformers into 15003-albert-tokenizer-special-tokens

sgugger commented:
If you have merged your branch with upstream then just a push should be enough to resolve the conflicts.

To rebase just do this on your branch:
```
git rebase origin/main
```
then you'll need to force-push
```
git push -u origin branch_name --force
```

beneyal committed:
Deberta tokenizer now behaves like Albert and its `convert_tokens_to_string` is not tested.

beneyal commented:
@sgugger Done 🙂 Whisper's test failed because an HTTP timeout, but the rest seem fine. Deberta V2 is now excluded from the sentencepiece tests, but has the same logic as Albert.

sgugger commented on tests/models/deberta_v2/test_tokenization_deberta_v2.py:
```diff
@@ -30,14 +30,14 @@ class DebertaV2TokenizationTest(TokenizerTesterMixin, unittest.TestCase):
     tokenizer_class = DebertaV2Tokenizer
     rust_tokenizer_class = DebertaV2TokenizerFast
-    test_sentencepiece = True
+    test_sentencepiece = False
```
I think this removes more tests that were passing. Might be better to just override the failing test to just pass, here.

beneyal replied:
Good idea, on it!

beneyal replied:
Pushed the fix. Please feel free to nit some more, I'm happy to work on this PR again! 🥳

sgugger reviewed (commented):
Almost there, one last nit and this should be ready for merge!

beneyal committed:
Ignore problematic tests in Deberta V2

sgugger reviewed (approved):
Thanks for iterating! Let's just add some comments on why those tests are skipped and we'll be good to merge!

beneyal committed:
Add comment on why the Deberta V2 tests are skipped

beneyal commented:
@sgugger Done, thanks for the suggestions!

sgugger commented:
Thanks again and sorry for the long wait on this PR!

ydshieh commented:
Hi @beneyal First, thank you for working on this. I have a question as we have a CI failure. With the following code snippet, 
```python
from transformers import T5Tokenizer, T5ForConditionalGeneration

tokenizer = T5Tokenizer.from_pretrained("t5-small")
model = T5ForConditionalGeneration.from_pretrained("t5-small")

input_ids = tokenizer("The <extra_id_0> walks in <extra_id_1> park", return_tensors="pt").input_ids

sequence_ids = model.generate(input_ids)
sequences = tokenizer.batch_decode(sequence_ids)
print(sequences)
```
we have different results:

**Before this PR**
```python
['<pad><extra_id_0> park offers<extra_id_1> the<extra_id_2> park.</s>']
```

**After this PR**
```python
['<pad> <extra_id_0> park offers<extra_id_1> the<extra_id_2> park.</s>']
```

Is this expected? 🙏

beneyal commented:
@ydshieh Glad to do it!

Yes, this is intended behavior, as it mimics the fast tokenizer:

```python
from transformers import T5Tokenizer, T5TokenizerFast, T5ForConditionalGeneration

tokenizer = T5Tokenizer.from_pretrained("t5-small")
tokenizer_fast = T5TokenizerFast.from_pretrained("t5-small")

model = T5ForConditionalGeneration.from_pretrained("t5-small")

input_ids = tokenizer("The <extra_id_0> walks in <extra_id_1> park", return_tensors="pt").input_ids
inputs_ids_fast = tokenizer_fast("The <extra_id_0> walks in <extra_id_1> park", return_tensors="pt").input_ids

assert input_ids.tolist() == inputs_ids_fast.tolist()

sequence_ids = model.generate(input_ids)

sequences = tokenizer.batch_decode(sequence_ids)
sequences_fast = tokenizer_fast.batch_decode(sequence_ids)

print(sequences, sequences == sequences_fast)  # outputs True
```
